Update tests to check if headnode is replaced when we update the SecurityGroups

himani2411 opened the pull request:
Check if the HeadNode instance is replaced if the Security Group ID and AdditionalSecurity Groups are updated.


### Checklist
- Make sure you are pointing to **the right branch**.
- If you're creating a patch for a branch other than `develop` add the branch name as prefix in the PR title (e\.g\. `[release-3.6]`).
- Check all commits' messages are clear, describing what and why vs how.
- Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Commit:
Add tests to check if headnode is replaced when we update the Security and Additional Security Groups

enrico-usai commented on tests/integration-tests/tests/networking/test_security_groups.py:
```diff
@@ -45,21 +45,37 @@ def test_additional_sg_and_ssh_from(region, custom_security_group, pcluster_conf
+
+    custom_security_group_stack = custom_security_group(count_of_sgs=count_of_sgs)
+    custom_security_group_ids = []
+    for sg_indx in range(count_of_sgs):
```
minor: let's use "sg_index" rather than "indx", it's just an extra character.

enrico-usai commented on tests/integration-tests/tests/networking/test_security_groups.py:
```diff
@@ -160,56 +187,65 @@ def _check_connections_between_head_node_and_compute_nodes(cluster):
+        for sg_indx in range(count_of_sgs):
+            security_group = template.add_resource(
+                SecurityGroup(
+                    f"SecurityGroupResource{sg_indx}",
```
Minor: Not added by your patch, but I'd remove the word "Resource" from the name. Everything we're creating is a resource.

enrico-usai commented on tests/integration-tests/tests/networking/test_security_groups.py:
```diff
@@ -160,56 +187,65 @@ def _check_connections_between_head_node_and_compute_nodes(cluster):
-                ToPort=65535,
-                CidrIp=vpc_cidr,
-                GroupId=Ref(security_group),
+def custom_security_group(vpc_stack, region, request):
```
minor: should be renamed to "custom_security_groups" because now is able to create more than 1 security group.

enrico-usai commented on tests/integration-tests/tests/networking/test_security_groups.py:
```diff
@@ -160,56 +187,65 @@ def _check_connections_between_head_node_and_compute_nodes(cluster):
+def custom_security_group(vpc_stack, region, request):
+    stacks_factory = CfnStacksFactory(request.config.getoption("credential"))
+
+    def _custom_security_group(count_of_sgs=1):
```
I'd rename the count_of_sgs to number_of_sgs, because it's not counting something, it's instead the number of SGs we're asking to create.

enrico-usai commented on tests/integration-tests/tests/networking/test_security_groups.py:
```diff
@@ -45,21 +45,37 @@ def test_additional_sg_and_ssh_from(region, custom_security_group, pcluster_conf
-    custom_security_group_id = custom_security_group.cfn_resources["SecurityGroupResource"]
+    count_of_sgs = 2
+
+    custom_security_group_stack = custom_security_group(count_of_sgs=count_of_sgs)
```
there is something strange on the fact that `custom_security_group` is returning a stack. 
I'd say this should return the list of security group ids instead, so we should move the logic to retrieve the ids into the custom_security_groups function.

In this way we'll encapsulate the management of the index into that function, and the caller doesn't need to know that the name of the resource is something in the form `"SecurityGroupResource{sg_indx}"`

himani2411 replied:
I followed the same modularity which was being followed earlier w.r.t `custom_security_group`. But then after changing the nos of outputs that it gives, it makes sense to return a list rather than a stack. Will make the change as you suggest

enrico-usai commented on tests/integration-tests/tests/networking/test_security_groups.py:
```diff
@@ -45,21 +45,37 @@ def test_additional_sg_and_ssh_from(region, custom_security_group, pcluster_conf
     fsx_name, efs_name = "fsx", "efs"
     cluster_config = pcluster_config_reader(
-        vpc_security_group_id=custom_security_group_id, fsx_name=fsx_name, efs_name=efs_name
+        vpc_security_group_id=custom_security_group_ids[0],
```
it's better to create a variable called for example `default_security_group = custom_security_group_ids[0]` and use it everywhere rather than using `custom_security_group_ids[0]` here and there.

enrico-usai commented on tests/integration-tests/tests/networking/test_security_groups.py:
```diff
@@ -45,21 +45,37 @@ def test_additional_sg_and_ssh_from(region, custom_security_group, pcluster_conf
+        vpc_security_group_id=custom_security_group_ids[0],
+        fsx_name=fsx_name,
+        efs_name=efs_name,
+        additional_security_groups=additional_security_groups,
```
This variable should be `assign_additional_security_groups`.

the boolean parameters should be "actions" in this way it's more clear the meaning of True/False.
Instead using `additional_security_groups` it seems the parameter is accepting a list of security groups.

enrico-usai commented on tests/integration-tests/tests/networking/test_security_groups.py:
```diff
@@ -106,19 +122,30 @@ def test_overwrite_sg(region, scheduler, custom_security_group, pcluster_config_
         logging.info("Checking SSH connection between cluster nodes before cluster update")
         _check_connections_between_head_node_and_compute_nodes(cluster)
+
+        # Update the Security Groups of HN to check if the HN instance is replaced.
```
let's use HeadNode rather than HN. AFAIK the HN acronym is not used in other places and Head Node is more self-explicative.

enrico-usai commented on tests/integration-tests/tests/networking/test_security_groups.py:
```diff
@@ -45,21 +45,37 @@ def test_additional_sg_and_ssh_from(region, custom_security_group, pcluster_conf
+        logging.info("Asserting that head node and compute node has and only has the custom security group")
+        assert_that(instances).is_length(3 if scheduler == "slurm" else 2)
+        for instance in instances:
+            assert_that(instance["SecurityGroups"]).is_length(1)
```
you could avoid the `if not additional_security_groups` by using `assert_that(instance["SecurityGroups"]).is_length(number_of_sgs)` in the check.

himani2411 replied:
I dont want to call the `_get_instances_by_security_group` if not necessary which is why I am using the `not additional_security_groups` in the check

enrico-usai commented on tests/integration-tests/tests/networking/test_security_groups.py:
```diff
@@ -106,19 +122,30 @@ def test_overwrite_sg(region, scheduler, custom_security_group, pcluster_config_
+            vpc_security_group_id=custom_security_group_ids,
+            additional_security_groups=additional_security_groups,
+        )
+        cluster.update(str(updated_config_file), force_update="true")
```
do we really need an extra update step? every update will take 15/30mins, I'd unify the changes in the `pcluster.config.update_sg.yaml` in the already existing `pcluster.config.update.yaml` file.

We can perform the update just one time and do all the checks we need.

himani2411 replied:
The reason for creating another yaml file itself, was because one of the checks in the `pcluster.config.update.yaml` is that the HeadNodes' managed sg allows access to Complte fleets custom SG

himani2411 replied:
Will make changes in the `test_additional_sg_and_ssh_from` tests for updating the cluster with a new sg_id for AdditionalSecurityGroups and SecurityGroups. However, this will not reduce the time for the tests a I would still need to update the cluster.

enrico-usai commented on tests/integration-tests/tests/networking/test_security_groups.py:
```diff
@@ -106,19 +122,30 @@ def test_overwrite_sg(region, scheduler, custom_security_group, pcluster_config_
+            additional_security_groups=additional_security_groups,
+        )
+        cluster.update(str(updated_config_file), force_update="true")
+        logging.info("Checking if the HeadNode is replaced after cluster update")
```
the message should be: `Verifying the HeadNode is not replaced after cluster update` otherwise it seems we're expecting it has to be replaced.

enrico-usai commented on tests/integration-tests/tests/networking/test_security_groups.py:
```diff
@@ -160,56 +187,65 @@ def _check_connections_between_head_node_and_compute_nodes(cluster):
 
     if not request.config.getoption("no_delete"):
-        cfn_stacks_factory.delete_stack(stack.name, region)
+        stacks_factory.delete_all_stacks()
```
do we really need a stack for each security group?
It's better to have the security group loop into the same CfnStack, they are logically related since used by the same tests and in this way it's easier to cleanup in case of failures/issues.

himani2411 replied:
Initially I was creating two stacks which I thought was excessive! 
But I later changed to creating one stack with multiple Security groups and missed the cleanup function here.
So we are creating a single stack but with multiple securityGroups.

enrico-usai commented on tests/integration-tests/tests/networking/test_security_groups/test_overwrite_sg/pcluster.config.update_sg.yaml:
```diff
@@ -0,0 +1,49 @@
+    {% if additional_security_groups == False %}
+    SecurityGroups:
+      - {{ vpc_security_group_id[1] }}
+    {% else %}
```
`SecurityGroups` and `AdditionalSecurityGroups` can be used together. 

So  if I'm not wrong we can have:
```
pcluster.config.yaml

  Networking:
        SecurityGroups:
          - {{ vpc_security_group_id[0] }}
```


```
pcluster.config.update.yaml


  Networking:
        SecurityGroups:
          - {{ vpc_security_group_id[1] }}
        AdditionalSecurityGroups:
          - {{ vpc_security_group_id[0] }}
```

In this way you can test at the same time what happens updating Security or AdditionalSecurityGroups.
feel free to find better alternatives anyway the goal is to avoid to execute 2 times the update command.

himani2411 replied:
I tried that initially. But apparently they both should not [co-exist](https://github.com/aws/aws-parallelcluster/blob/0b79cf32a1bc01d19a4295e8592a8b9cd4ae9dc7/cli/src/pcluster/schemas/cluster_schema.py#L653-L654)
https://github.com/aws/aws-parallelcluster/blob/0b79cf32a1bc01d19a4295e8592a8b9cd4ae9dc7/cli/src/pcluster/schemas/cluster_schema.py#L653-L654

himani2411 commented on tests/integration-tests/tests/networking/test_security_groups.py:
```diff
@@ -160,56 +190,66 @@
-    )
-    cfn_stacks_factory.create_stack(stack)
+        stacks_factory.create_stack(stack)
+        sg_stack_name = stack.name
```
Working on this!

Commit:
Add --custom-security-groups-stack-name integ test option.
Add the Headnode replacement verification in test_additional_sg_and_ssh_from function

enrico-usai commented on tests/integration-tests/tests/networking/test_security_groups.py:
```diff
@@ -14,21 +14,21 @@
     The additional security group should be added to the head and compute nodes. The
     """
-    custom_security_group_id = custom_security_group.cfn_resources["SecurityGroupResource"]
+    custom_security_group_id = custom_security_groups
```
`custom_security_groups` is always returning a list now, so this should be:
```python
custom_security_group_id = custom_security_groups[0]
```

himani2411 replied:
I just made the change!

himani2411 committed:
Merge branch 'develop' into wip/integ-tests
